client/sftp_client.py

```diff
@@ -196,202 +614,749 @@ def load_dirs(self, path, parent=None):
+	### 	group, owner, or other then perform bitshifts on the mode
 	def _check_perm(self, fullname):
-		lstatout = self.ftp.stat(fullname)
-		mode = lstatout.st_mode
+		mode = self.stat(fullname).st_mode
+		uid = self.stat(fullname).st_uid
+		gid = self.stat(fullname).st_gid
+		owner = False
+		group = False
+		other = False
+		try:
+			self.ftp.chmod(fullname, mode)
+		except IOError as e:
+			if e[0] == 13:
+				other = True
+			else:
+				raise e
+		else:
+			try:
+				self.ftp.chown(fullname, uid, gid)
+			except IOError as e:
+				if e[0] == 13:
+					group = True
+				else:
+					raise
+			else:
+				owner = True
 		perm = '   '
-		_file = self.ftp.file(fullname)
-		perm += 'r' if _file.readable() else '-'
-		perm += 'w' if _file.writable() else '-'
-		_file.close()
-		perm += 'x' if bool(int(oct(mode)[2:]) & stat.S_IXOTH) else '-'
+		if other:
+			perm += 'r' if bool(mode & stat.S_IROTH) else '-'
+			perm += 'w' if bool(mode & stat.S_IWOTH) else '-'
+			perm += 'x' if bool(mode & stat.S_IXOTH) else '-'
+		elif group:
+			perm += 'r' if bool(mode & stat.S_IRGRP) else '-'
+			perm += 'w' if bool(mode & stat.S_IWGRP) else '-'
+			perm += 'x' if bool(mode & stat.S_IXGRP) else '-'
+		elif owner:
+			perm += 'r' if bool(mode & stat.S_IRUSR) else '-'
+			perm += 'w' if bool(mode & stat.S_IWUSR) else '-'
+			perm += 'x' if bool(mode & stat.S_IXUSR) else '-'
+		else:
+			raise ValueError('Permission not parsed correctly')
 		return perm
+	###FIX ME
 
 	def _yield_dir_list(self, path):
 		for name in self.ftp.listdir(path):
 			yield name
 
+	def _already_exists(self, path):
+		try:
+			self.ftp.stat(path)
+		except IOError as error:
+			if error[0] == 2:
+				return False
+			else:
+				 raise error
+		else:
+			return True
+
+	def _rename_file(self, _iter, path):
+		self.ftp.rename(self._tv_model[_iter][2], path)  #pylint: disable=unsubscriptable-object
+
+	@handle_permission_denied
+	def _make_file(self, path, ftp=None):
+		# If method called when self.ftp is busy, alternate ftp is specified, prevents Garbage Packet Error
+		if ftp is None:
+			ftp = self.ftp
+		ftp.mkdir(path)
+
 	def _get_raw_time(self, fullname):
-		lstatout = self.ftp.lstat(fullname)
+		lstatout = self.ftp.stat(fullname)
 		return lstatout.st_mtime
 
-	def _get_is_folder(self, fullname):
-		lstatout = self.ftp.lstat(fullname)
-		mode = lstatout.st_mode
-		return stat.S_ISDIR(mode)
+	@handle_permission_denied
+	def delete(self, model, treeiter):
+		name = self._tv_model[treeiter][2]
+		if self.get_is_folder(name):  #pylint: disable=unsubscriptable-object
+			if not self.remove_by_folder_name(name):
+				return
+		else:
+			if not self.remove_by_file_name(name):
+				return
+		gui_utilities.show_dialog_warning('Successfully deleted ' + name, self.application.get_active_window())
+		self.refresh()
+		if self.parentless:
+			self._tv_model.clear()
+			self.load_dirs(self.default_directory)
+
+	###FIX ME
```

Comment:
What exactly needs to be fixed here? If there's something outstanding it needs to be documented.